libs/runtime/unit/platform_unit_test.cpp

```diff
@@ -1070,39 +1076,140 @@ TEST_CASE("get_authentication_id", "[platform]")
+    // See https://en.wikipedia.org/wiki/Chi-squared_test for details.
+    // The chi-squared statistic is the sum of the squared difference between the observed and expected values
+    // divided by the expected value. If the chi-squared statistic is less than the critical value, the hypothesis
+    // is accepted.
+
+    double zero_count = 0;
+    double one_count = 0;
+    double expected_value = static_cast<double>(sequence_length) * sizeof(uint32_t) * 8 / 2;
+
+    // Treat each bit in the random number as degree of freedom.
+    for (size_t i = 0; i < sequence_length; i++) {
+        unsigned long value = static_cast<int>(random_number_generator());
+        size_t bit_count = __popcnt(value);
+        zero_count += static_cast<double>(32 - bit_count);
+        one_count += static_cast<double>(bit_count);
     }
 
-    double chi_squared_statistic = 0.0;
-    for (int i = 0; i < NUM_BINS; i++) {
-        double observed = static_cast<double>(observed_values[i]);
-        chi_squared_statistic += pow(observed - expected_value, 2) / expected_value;
-    }
+    double chi_squared_statistic = std::pow(zero_count - expected_value, 2) / expected_value;
+    chi_squared_statistic += std::pow(one_count - expected_value, 2) / expected_value;
 
     double critical_value = CHI_SQUARED_STATISTIC_THRESHOLD;
     std::cout << chi_squared_statistic << std::endl;
     return chi_squared_statistic < critical_value;
 }
 
+typedef std::complex<double> Complex;
+#if !defined(M_PI)
+#define M_PI (3.14159265358979323846264338327950288)
+#endif
+
+/**
+ * @brief Perform a Fast Fourier Transform on the input sequence.
+ * An implementation of "Cooley-Tukey Radix-2 Decimation in Time"
+ * See: https://en.wikipedia.org/wiki/Cooley%E2%80%93Tukey_FFT_algorithm
+ *
+ * @param[in,out] samples The input sequence.  On output, the FFT of the input sequence.
+ * @param[in] invert If true, perform an inverse FFT.
+ */
+void
+fft(std::vector<Complex>& samples, bool invert = false)
+{
+    size_t n = samples.size();
+    if (n <= 1)
+        return;
+
+    std::vector<Complex> even(n / 2), odd(n / 2);
+    for (size_t i = 0, j = 0; i < n; i += 2, ++j) {
+        even[j] = samples[i];
+        odd[j] = samples[i + 1];
+    }
+
+    fft(even, invert);
+    fft(odd, invert);
+
+    double angle = 2 * M_PI / n * (invert ? -1 : 1);
+    Complex w(1), wn(std::cos(angle), std::sin(angle));
+
+    for (size_t i = 0; i < n / 2; ++i) {
+        Complex t = w * odd[i];
+        samples[i] = even[i] + t;
+        samples[i + n / 2] = even[i] - t;
+        if (invert) {
+            samples[i] /= 2;
+            samples[i + n / 2] /= 2;
+        }
+        w *= wn;
+    }
+}
+
+/**
+ * @brief Determine the provided random number generator has a dominant frequency in its output.
+ *
+ * @param[in] sequence_length The number of random numbers to examine. Must be a power of 2.
```

Comment:
This line says it must be a power of 2 but i don't see any code in the function that checks this or relies on this.

Reply:
The fft function is recursive and divides the samples into 2 (even and odd) and recursively calls fft on each odd and even sequence. I will add a check for power of 2.